rc/windowing/detection.kak

```diff
@@ -0,0 +1,55 @@
+# Attempt to detect the windowing environment we're operating in
+#
+# Supported environments and their defaults are:
+#  * X11 - each new window is a new top level window
+#  * tmux - new windows are vertical tmux panes
+#  * screen - new windows are vertical screen regions
+#  * iterm - new windows are vertical iterm panes
+#  * kitty - new windows are kitty windows
+#
+# Each module must define a detection command "detect-module-<mod>" which
+# adds its module name to the str-list option "windowing_modules_detected" if
+# it determines that it is relevant to the current runtime environment.
+#
+# Exactly one module will be loaded from the list of detected modules. The
+# "windowing_modules" list specifies the list of modules to search for, and
+# the order of priority for loading.
+#
+# Each module must define two aliases:
+#  * terminal - create a new terminal with sensible defaults
+#  * focus - focus the specified client, defaulting to the current client
+#
+
+declare-option str-list windowing_modules_detected
+declare-option str-list windowing_modules 'tmux' 'screen' 'kitty' 'iterm' 'x11'
+
+define-command -hidden detect-windowing-modules -docstring '
+detect-windowing-module: run the detection functions for the windowing modules.
+This creates the detected_modules str-list, containing only the modules
+relevant to the current runtime environment.' \
+%{
+    evaluate-commands %sh{
+        wmods=${kak_opt_windowing_modules}
+        for mod in ${wmods}; do
+            echo "try %{ detect-module-${mod} }"
+        done
+    }
+}
+
+hook -group windowing global KakBegin .* %{
+
+    # detect the windowing modules relevant to the current environment
+    detect-windowing-modules
+
+    # load the first module on the list.
+    #
+    # At this point we know that these modules exist and are relevant. The list
+    # is ordered by preference, so we can simply load the first module on the
+    # list.
+    evaluate-commands %sh{
+        for i in ${kak_opt_windowing_modules_detected}; do
+            echo "require-module ${i}"
+            break
+        done
```

Comment:
We probably don't need a loop here:

```sh
printf 'require-module %s' "${kak_opt_windowing_modules_detected%% *}"
```

Reply:
Nice - I was stumped by the lack of arrays (I'm totally a bash person), but I didn't think of that approach.

Reply:
To also support the scenario where a module is not included by the user, we can try the next one:

```sh
define-command require-windowing-module %{ evaluate-commands %sh{
	set -- $kak_opt_windowing_modules_detected
	module="$1"
	test -n "$module" || exit # No windowing module available.
	shift
	printf %s "try %{
		# Try to load the first module.
		require-module $1
	} catch {
		# If the module failed to load, remove it from the list of modules and call
		# ourselves recursively.
		set-option global windowing_modules_detected $@
		require-windowing-module
	}"
} }
```

This is just like described in https://github.com/mawww/kakoune/pull/3457#issuecomment-636394625, but without the `set-option -remove` feature, which would make this a bit nicer.

Reply:
Is that kind of recursion supported in this context? I'm not sure exactly how the %sh{} construct is implemented, but if each one is its own /bin/sh interpreter there's scope for this to blow up pretty seriously.

Also, is that a better approach than using a single additional option to collect state for further processing? It seems like a far more complex solution in many ways, even aside from the recursion question.

Reply:
Also, reimplemented so it does the recursion within a single shell invocation (untested, just trying this idea on for size):

```
define-command require-windowing-module %{ evaluate-commands %sh{
    try_require_module () {
        set -- $kak_opt_windowing_modules_detected
        if test -n "$1"; do
            module=$1
            shift
            printf %s "try %{
                # Try to load the first module.
                require-module $module
                set current windowing_modules_detected
            } catch {
                set current windowing_modules_detected $@
            }"
            try_require_module
        fi
    }
    try_require_module
} }
```
I'm still not sure it's a better approach than my first one, but it avoids recursing with each level being a whole new shell.

Reply:
@krobelus this recursive model works for loading the first module in the list, but that's only useful if we already have a list of modules that's been filtered to remove anything not relevant to the current environment. And once you've got that filtered list you don't need to do this kind of recursive jiggery-pokery, you can just load the first module in the list.

I've tried implementing something that replaces the `require-module` with a call to a function provided by each module that checks its environment and calls `require-module` if it's relevant, but it just doesn't work. I'm not sure why it's failing - the functions are very simple, and the only difference between this version (which runs fine, loading the first module in the list) and the failing version is replacing `require-module $module` with a call to `try-require-module-$module`, defined something like this:
```
define-command -hidden try-require-module-tmux %{
    evaluate-commands %sh{
        if [ -n "$TMUX" ]; then
            printf %s "require-module tmux"
        else
            printf %s "fail 'Could not find tmux environment'"
        fi
    }
}
```
I'm really not sure why this isn't working, but even if it did I think it's a lot more convoluted than the two-step approach in the current patch set.